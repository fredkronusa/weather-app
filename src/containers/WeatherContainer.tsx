import { ICurrentForecast, IWeeklyForecast } from "../interfaces/Weather";
import React, { useEffect, useState } from "react";
import {
  createCurrentForecastData,
  createWeeklyForecastData,
} from "../utils/utils";
import {
  getCurrentForecastUrl,
  getWeeklyForecastUrl,
  initialCurrentForecast,
  initialWeeklyForecast,
} from "../config/config";

import CurrentForecastComponent from "../components/CurrentForecast";
import LocationSearchComponent from "../components/LocationSearch";
import MetricsSwitchComponent from "../components/MetricsSwitch";
import WeeklyForecastComponent from "../components/WeeklyForecast";
import { colors } from "../common/styles";
import styled from "styled-components";

const HeaderWrapper = styled.div`
  background: ${colors.lightGrey};
  display: flex;
  justify-content: space-between;
  padding: 20px 40px;
  align-items: center;
`;

const Spinner = styled.img`
  margin: 0 auto;
  display: block;
`;

const WeatherContainer = () => {
  const [currentForecast, setCurrentForecast] = useState<ICurrentForecast>(
    initialCurrentForecast
  );
  const [weeklyForecast, setWeeklyForecast] = useState<IWeeklyForecast>(
    initialWeeklyForecast
  );

  const [isCelsius, setIsCelsius] = useState<boolean>(true);
  const [loading, setLoading] = useState<boolean>(true);
  const [location, setLocation] = useState<string>("Melbourne");

  useEffect(() => {
    setLoading(true);

    const getWeeklyForecastData = async () => {
      const weatherInfo = await fetch(getWeeklyForecastUrl(location));
      const response = await weatherInfo.json();
      setWeeklyForecast(createWeeklyForecastData(response));
    };

    const getCurrentForecastData = async () => {
      const weatherInfo = await fetch(getCurrentForecastUrl(location));
      const response = await weatherInfo.json();
      setCurrentForecast(createCurrentForecastData(response.data[0]));
    };

    getWeeklyForecastData();
    getCurrentForecastData();
  }, [location]);

  useEffect(() => {
    if (weeklyForecast.weeklyForecast.length && currentForecast.city) {
      setLoading(false);
    }
  }, [weeklyForecast, currentForecast]);

  return (
    <div>
      <HeaderWrapper>
        <LocationSearchComponent
          location={location}
          setLocation={setLocation}
        />
        <MetricsSwitchComponent setIsCelsius={setIsCelsius} />
      </HeaderWrapper>

      {!loading ? (
        <>
          <CurrentForecastComponent
            currentForecast={currentForecast}
            isCelsius={isCelsius}
          />
          <WeeklyForecastComponent
            weeklyForecast={weeklyForecast.weeklyForecast}
            isCelsius={isCelsius}
          />
        </>
      ) : (
        <Spinner src={"images/spinner.gif"} alt="loading" />
      )}
    </div>
  );
};

export default WeatherContainer;
