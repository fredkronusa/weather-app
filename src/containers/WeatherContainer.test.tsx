import * as React from "react";

import WeatherContainer from "./WeatherContainer";
import { render } from "@testing-library/react";

describe("<WeatherContainer />", () => {
  // SNAPSHOT
  it("should match snapshot", async () => {
    const { asFragment } = render(<WeatherContainer />);

    expect(asFragment()).toMatchSnapshot();
  });
});
