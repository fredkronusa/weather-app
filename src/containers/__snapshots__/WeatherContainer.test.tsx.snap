// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherContainer /> should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-fzoLag kSlmIv"
    >
      <div>
        <label
          class="sc-AxmLO fGmgQu"
          for="citySearch"
        >
          Search Weather
        </label>
        <input
          class="sc-fzozJi ffhvoi"
          id="citySearch"
          name="citySearch"
          type="text"
          value="Melbourne"
        />
        <input
          class="sc-fzoLsD hNoulZ"
          type="button"
          value="GO"
        />
      </div>
      <div>
        <input
          checked=""
          class="sc-fzplWN caKVrL"
          id="celsius"
          name="unitType"
          type="radio"
          value="celsius"
        />
        <label
          class="sc-fzpans dFlUAH"
          for="celsius"
        >
          ℃
        </label>
        <input
          class="sc-fzplWN caKVrL"
          id="fahrenheit"
          name="unitType"
          type="radio"
          value="fahrenheit"
        />
        <label
          class="sc-fzpans dFlUAH"
          for="fahrenheit"
        >
          ℉
        </label>
      </div>
    </div>
    <img
      alt="loading"
      class="sc-fzoXzr kiyFbV"
      src="images/spinner.gif"
    />
  </div>
</DocumentFragment>
`;
